streaming/src/main/scala/org/apache/spark/streaming/receiver/CongestionStrategyImpl.scala

```diff
@@ -0,0 +1,102 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.receiver
+
+import scala.collection.mutable.ArrayBuffer
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.Random
+import org.apache.spark.util.random.{RandomSampler, BernoulliSampler}
+
+/**
+ * This class provides a congestion strategy that ignores
+ * any back-pressure information.
+ * @see CongestionStrategy
+ */
+class IgnoreCongestionStrategy extends CongestionStrategy {
+
+  override def onBlockBoundUpdate(bound: Int): Unit = {}
+
+  override def restrictCurrentBuffer(currentBuffer: ArrayBuffer[Any],
+                                     nextBuffer: ArrayBuffer[Any]): Unit = {}
+}
+
+class PushBackCongestionStrategy(blockInterval: Long) extends CongestionStrategy {
+
+  private val latestBound = new AtomicInteger(-1)
+
+  override def onBlockBoundUpdate(bound: Int): Unit = latestBound.set(bound)
+
+  override def restrictCurrentBuffer(currentBuffer: ArrayBuffer[Any],
+                                     nextBuffer: ArrayBuffer[Any]): Unit = {
+    val bound = latestBound.get()
+    val difference = currentBuffer.size - bound
+    if (bound > 0 && difference > 0) {
+      nextBuffer ++=: currentBuffer.takeRight(difference)
+      currentBuffer.reduceToSize(bound)
+    }
+    // We've had our fill for the amount of time it would take us to process the difference.
+    // Use the fact this is synchronized with data ingestion to prevent more data coming in.
+    // We need to wait for quite less than the block interval for the next 'clock tick'
+    // (and block generation) to occur normally : this only leaves 5% of the block interval
+    // during which data can 'sneak in'
+    val delay = math.round( blockInterval * math.min((difference.toFloat / bound), 0.9) )
+    Thread.sleep(delay)
+  }
+
+}
+
+class DropCongestionStrategy extends CongestionStrategy {
+
+  private val latestBound = new AtomicInteger(-1)
+
+  override def onBlockBoundUpdate(bound: Int): Unit = latestBound.set(bound)
+
+  override def restrictCurrentBuffer(currentBuffer: ArrayBuffer[Any],
+                                     nextBuffer: ArrayBuffer[Any]): Unit = {
+    val bound = latestBound.get()
+    val difference = currentBuffer.size - bound
+    if (bound > 0 && difference > 0) {
+      currentBuffer.reduceToSize(bound)
+    }
+  }
+
+}
+
+class SamplingCongestionStrategy extends CongestionStrategy {
+
+  private val rng = RandomSampler.newDefaultRNG
+
+  private val latestBound = new AtomicInteger(-1)
+
+  override def onBlockBoundUpdate(bound: Int): Unit = latestBound.set(bound)
+
+  override def restrictCurrentBuffer(currentBuffer: ArrayBuffer[Any],
+                                     nextBuffer: ArrayBuffer[Any]): Unit = {
+    val bound = latestBound.get()
+    val f = bound.toDouble / currentBuffer.size
+    if (f > 0 && f < 1){
+      val samplees = currentBuffer.to
```

Comment:
OK! So, the `currentBuffer.toIterator` is lazy, so is `BernouilliSampler()`, so is `sample()`. So that, if I don't make a copy of the buffer once (in `samplees`), by the time I call `currentBuffer.clear()`, I've lost all elements in `sampled`.

I refactored this part to make things clearer (and copy *after* sampling rather than before) in 92767f5 (the copy is made in the `toList`).